pkg/runtime/runtime.go

```diff
@@ -758,6 +780,45 @@ func (a *DaprRuntime) initPubSub() error {
 	return nil
 }
 
+// Publish is an adapter method for the runtime to pre-validate publish requests
+// And then forward them to the Pub/Sub component.
+// This method is used by the HTTP and gRPC APIs.
+func (a *DaprRuntime) Publish(req *pubsub.PublishRequest) error {
+	if allowed := a.isPubSubOperationAllowed(req.Topic, a.allowedPublishings); !allowed {
+		return fmt.Errorf("topic %s is not allowed for app id %s", req.Topic, a.runtimeConfig.ID)
+	}
+	return a.pubSub.Publish(req)
+}
+
+func (a *DaprRuntime) isPubSubOperationAllowed(topic string, topicsList []string) bool {
+	allowed := false
+
+	// first check if allowedTopics contain it
+	if len(a.allowedTopics) > 0 {
+		for _, t := range a.allowedTopics {
+			if t == topic {
+				allowed = true
+				break
+			}
+		}
+		if !allowed {
+			return allowed
+		}
+	}
+
+	// check if a granular scope has been applied
+	allowed = false
+	if len(topicsList) > 0 {
```

Comment:
possibly we can refactor out this loop finding logic and use across...

Reply:
use across.. what?

Reply:
> use across.. what?

I meant refactor this common finding logic in loop and call from places in runtime.go if possible...

Reply:
It's already refactored inside `isPubSubOperationAllowed`, and called from multiple places..